rfcs/003-nix-nickel.md

```diff
@@ -0,0 +1,460 @@
+in the PARM, just for the sake of consistency.
+
+**TODO**: clarify what this last sentence exactly means.
+
+### Nickel and Nixpkgs
+
+As underlined in [Interaction with Nixpkgs](#interaction-with-nixpkgs), an FFI
+mechanism would require a bidirectional communication between Nix and Nickel.
+This is made complicated by the nature of both languages, which are lazy and
+functional, meaning that a Nix function called from Nickel may incur arbitrary
+calls back to Nickel (and vice-versa) to apply functions or to evaluate thunks.
+Not only does this seem complex, but also hard to be made efficient.
+
+Another possibility is to have everything evaluate in only one world, either Nix
+or Nickel. Practically, we could compile the Nix code to Nickel or vice-versa,
+and run everything on one side only.
+
+Compiling Nickel to Nix and run everything on the Nix side would have the
+advantage of leveraging all the Nix primitives and machinery freely. However,
+this solution simply nullifies a lot of the value proposition of Nickel. The
+added value of contracts is mainly due to its careful design as a primitive
+feature of the language, and while it could be simulated to some extent in Nix,
+the error message and performances just couldn't be in par. Similarly, the merge
+system being native is in part motivated by the perspective of better
+performance and error message. This simply becomes moot if we compile those
+features away back to plain Nix code.
+
+On the other hand, compiling Nix to Nickel looks more appealing. Nickel is close
+to being a superset of Nix (the only non-trivial missing features being the
+`with` construct and string contexts, but the latter doesn't even really change
+the semantics). This would preserve the Nickel advances, and we could leverage
+directly any non-builtin Nix function, including all the Nixpkgs library
+functions.
+
+The potential drawbacks of this approach:
+
+- Performance: Nickel is not expected to be in par with Nix at this point, so
+  the performance of Nixpkgs compiled to Nickel might be prohibitively bad at
+  first. On a positive note, this would provide a great benchmark suite and
+  motivation to guide future performance improvements.
+- Some builtin may incur effects, which are not trivial to combine with the
+  execution model (revertible thunks). That being said, those effects are in
+  practice idempotent and commutative (creating paths in the store, creating
+  derivation, etc.), which should be fine.
+
+Compilation could be done on the fly, as the compilation process would be rather
+straightforward, or have a Nixpkgs snapshot pre-compiled to Nickel, or a mix of
+both.
+
+This solution requires to reimplement Nix builtins in Nickel (the nickel-nix
+compatibility layer) . This is an interesting milestone in itself, because even
+without a Nix-to-Nickel compiler, the compatibility layer would already make
+writing derivations in pure Nickel possible.
```

Comment:
If it's a lot of work to reimplement builtins, I was wondering whether it may be possible to leverage the builtins within Nix itself (i.e. the executable / C++ source code) by using Nickel as a library in Nix. I'm not familiar with the Nix implementation, so it may be cumbersome, as often extension functions are tied closely to the interpreter. However, with some refactoring, a shared interface may be extracted for use in Nix proper and Nix-Nickel.

Using Nickel as a library seems like a good path for Nix in the future, whatever happens. Once that's in place, it makes sense to slowly port Nix from C++ to Rust. I believe there is already some interest in that direction.

Reply:
Yes, that is a possible path forward, but it is in opposition with the constraint `_Do not require unreasonable changes to Nix itself_`. We could debate if those changes are unreasonable, but I think that for a first approach, the less changes to Nix, the better. Not only from a disturbance/community point of view, but also I fear that even once accepted, the Nix side of the changes would be very long to land given the current development process, and that would hold us back.